### PR TITLE
Don't stop on the first error

### PR DESCRIPTION
**Pull request by diml**
Currently jbuilder stops as soon as an error is encountered. After this patch, jbuilder continues until nothing more can be built. This is essentially the same as with `make -k`.

## Implementation

The core of this work is to change the semantic of `Future` so that we can wait for all started lightweight threads to terminate. I initially tried this as a layer on top of `Future`, but I was finding it hard to reason about background computations using the `Lwt`/`Async` semantic. So instead I rewrote `Future` as ` Fiber` module with a much simpler semantic. Additionally, the implementation is simpler and I expect that linear code will run faster and allocate less as well.
 
The main differences with `Future` are the following:

- a given `Fiber.t` must be used only once. Using it twice will result in it the computation it describes being executed twice. One must use a `Fiber.Ivar.t` in order to share the result of a fiber
- all fibers must be connected to the scheduler, i.e. writing this has no effect:
  ```ocaml
  ignore (Process.run ... : unit Fiber.t)
  ```
- the only way to introduce concurrency is via the primitives such as `Fiber.fork`, i.e. this doesn't introduce concurrency:
  ```ocaml
  let a = Process.run ... in
  let b = Process.run ... in
  a >>= fun x ->
  b >>= fun y ->
  ...
  ```
- we have much stronger guarantees about when background errors can occur and when the execution of a fiber has fully completed. For instance it is possible to collect all errors raised during the execution of a fiber and be sure that the fiber will make no further progress after that, i.e. that no background computations are left over

### CODE REVIEW

**Comment by lpw25** [comment]
The fiber interface looks good. For the `fork` functions:
```ocaml
val fork : (unit -> 'a t) -> (unit -> 'b t) -> ('a * 'b) t

val nfork_map  : 'a list -> f:('a -> 'b   t) -> 'b list t
```
I believe a more traditional interface would be something like:
```ocaml
val fork : (unit -> 'a t) -> 'a Future.t t

val nfork : (unit -> 'a t) list -> 'a Future.t list t
```
So that instead of:
```
let foo () =
  fork bar (fun () -> ...do stuff in parallel...) >>= fun (bar, stuff) ->
  ...do stuff when both are finshed...
```
you write something like:
```
let foo () =
  fork bar >>= fun fut ->
  ...do stuff in parallel... >>= fun stuff ->
  Future.wait fut >>= fun bar ->
  ...do stuff when both are finished...
```
Where a `Future.t` is just an `Ivar.t` that can also hold an exception as a result, so that if `bar` raises an exception then you can reraise it inside `foo`.

**Comment by diml** [comment]
> I believe a more traditional interface would be something like:

Yh that's more generic indeed.

> val both : 'a t -> 'b t -> ('a * 'b) t

I initially started with this type, i.e. without the `unit -> ...` but there are many places where constructing the `_ t` might raise, so I think it's safer to put the `unit -> ...` here, even if the resulting API is not as nice.

> Where a Future.t is just an Ivar.t that can also hold an exception as a result, so that if bar raises an exception then you can reraise it inside foo.

I'm not sure reraising the exception fits well with the current way errors are handled. Something that would fit better would be saying that the `Future.t` never gets determined if the execution of the fiber raises. Then one could combine fork with existing functions in order to cactch errors:

```
Future.fork (fun () -> wait_errors (fun () -> bar)) >>= fun fut ->
...
Future.wait fut >>= function
| Ok x -> ...
| Error () -> raise Already_seen
```

**Comment by diml** [comment]
I'm not sure about the name `wait_errors` BTW. What it does is wait for all computations started by the fibers to terminate and then return whether the execution succeeded or whether at least one exception was raised.

**Comment by lpw25** [comment]
> val both : 'a t -> 'b t -> ('a * 'b) t
>
> I initially started with this type, i.e. without the unit -> ... but there are many places where constructing the _ t might raise, so I think it's safer to put the unit -> ... here, even if the resulting API is not as nice.

But you were defining a function that gave concurrency, I just mean a function that is equivalent to:

```ocaml
fun a b ->
  a >>= fun a ->
  b >>| fun b ->
    (a, b)
```

but with a more efficient implementation (since you statically know what that bind is going to do). If you have `fork` then I don't think you need `both` to produce concurrency, and then I don't think you need to worry as much about exceptions around it.

> I'm not sure reraising the exception fits well with the current way errors are handled. Something that would fit better would be saying that the Future.t never gets determined if the execution of the fiber raises.

I see, you're thinking of the semantics where exceptions are propagated through `fork` rather than through `wait`. So that running:

```ocaml
fork (fun () -> error "Foo") >>= fun _unused_future ->
return 3
```
will essentially return a pair of `[Error "Foo"]` and `3`, whereas the semantics I was proposing would drop that exception since nothing depended on the resulting future. That semantics is fine too.

As a side note, whilst you clearly need a semantics for what to do with exceptions for genuine fatal errors, it might be better to provide a separate `error` effect for dealing with normal errors, and include the error type in the type of the computation. So:

```ocaml
type ('a, 'error) t

val return : 'a -> ('a, 'error) t
val bind : 'a t -> ('a -> 'b t) -> 'b t
val error : 'error -> ('a, 'error) t
```

**Comment by diml** [comment]
> I see, you're thinking of the semantics where exceptions are propagated through fork rather than through wait

Indeed. I find this semantic easier to reason about since you are guaranteed that all exceptions will be seen by an error handler.

> As a side note, whilst you clearly need a semantics for what to do with exceptions for genuine fatal errors, it might be better to provide a separate error effect for dealing with normal errors, and include the error type in the type of the computation. So:

What is the advantage of this over using `('a, 'error) result Fiber.t` wherever you need it?

**Comment by lpw25** [comment]
> What is the advantage of this over using ('a, 'error) result Fiber.t wherever you need it?

The automatic propagation and collection of errors. I'm suggesting they get handled the way you are suggesting that exceptions get handled. So `fork` would gather all the errors, and something like `wait_errors` would let you execute all sub-fibers to completion and see their errors.

Note that we were thinking of doing a very similar thing for Tenacious in jenga for collecting all messages -- not just errors. This becomes particularly useful once "polling mode" forces you to add backtracking to the monad because it takes care of the business of backtracking the messages as well.

**Comment by diml** [comment]
I added the `fork` function you suggested and renamed the old ones `fork_and_join`, `parallel_map`, ... It's now looking quite nice. I also moved the fiber module to its own library.

File: src/report_error.mli [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    Because jbuilder doesn't stop at the first error, it might end up reporting the same
+    error twice about missing libraries for instance. To avoid this, we keep a cache of
+    reported errors and ignore errors that have already been reported.
+
+    We cache what is actually printed to the screen.
+*)
+val report : exn -> unit
+
+(**/**)
+val map_fname : (string -> string) ref
```
Comment (rgrinberg) [inline src/report_error.mli]: Is this `map_fname` set anywhere? Also, maybe a high level comment the purpose?

File: src/super_context.ml [hunk 2]
```diff
@@ -10,7 +10,7 @@ module Dir_with_jbuild = struct
     { src_dir : Path.t
     ; ctx_dir : Path.t
     ; stanzas : Stanzas.t
-    ; scope   : Lib_db.Scope.t Lib_db.with_required_by
+    ; scope   : Lib_db.Scope.t With_required_by.t
```
Comment (rgrinberg) [inline src/super_context.ml]: I was eyeing at a change like this as well. I'm wondering why is a part of this PR though, I don't quite see the dependency? Can we split this into its own PR or things are a bit too messy.

File: src/report_error.mli (hunk 1 above)
Comment (diml) [reply src/report_error.mli]: That something we use at Jane Street for the public release process. Basically during the process files and we use this function to map errors to the original files rather than the exported projects. I've been meaning to change the code to just read the output of `jbuilder external-lib-deps` and use a regexp to change filenames so that this can be removed, but I haven't done it yet

Comment (diml) [reply src/report_error.mli]: so eventually this will disappear. It's kind of ugly...

File: src/super_context.ml (hunk 2 above)
Comment (diml) [reply src/super_context.ml]: It's because I wanted `With_required_by.reraise`, and then it felt natural to put this type in there. It's true that this PR is big. I can try to split it

**Comment by diml** [comment]
Acutally, it's going to be painful to split. I should have created smaller features right from the start. I'm eager to merge this as once you start using it it's hard to go back.

Here is a sum up of changes in this feature:
- addition of a `With_required_by` module. The type `Lib_db.with_required_by` was moved there and it also allow add a dependency stack to an exception
- replaced various strings by `With_required_by.Entry.t`, which represent an entry in a dependency stack. We use a constructed type to delay the printing to a string
- used `With_required_by.Entry.t` to represent dependency stacks in `Findlib`
- moved error reporting from `Main` to a new module `Report_error`. In particular, this module keeps of what has been shown to the user, to avoid showing several time the same error
- added an `Errors` module to collect the various exceptions in the code
- moved process management from `Future` to a new module `Process`
- moved the scheduler from `Future` to a new module `Scheduler`
- deleted the rest of `Future` and replaced it by a new module `Fiber` in its own library
- added a `result` library, which is shared between `fiber` and `cmdliner`

Eventually, I'd like to release the fiber library. I'm not sure whether it should be a library on its own that we vendor in Dune or whether it should just be distributed with Dune.

**Comment by rgrinberg** [comment]
Ok, thanks. The summary is useful. Fibers as a lib can still be totally
independent even if it's part of the dune repo
